L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -150,92 +150,93 @@ void HGCalClusteringImpl::NNKernel( std::vector<edm::Ptr<l1t::HGCalTriggerCell>>
+        // as well as all neighbor TC
         std::unordered_set<uint32_t> cluTcSet;
-
-        const edm::PtrVector<l1t::HGCalTriggerCell> pertinentTC_clu1 = clustersTmp.at(i_clu1).triggercells();
-        
-        for(edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc_clu1 = pertinentTC_clu1.begin(); tc_clu1 != pertinentTC_clu1.end(); ++tc_clu1 ){
-            cluTcSet.insert( (*tc_clu1)->detId() );
-            HGCalDetId tcInClu1DetId( (*tc_clu1)->detId() );
-            const auto neighbors = triggerGeometry.getNeighborsFromTriggerCell( tcInClu1DetId );
-            for(const auto neighbor : neighbors)
-            {
-                if( !( cluTcSet.find( neighbor ) == cluTcSet.end() ) )
-                {                     
-                    cluTcSet.insert( neighbor );
-                }
+        for(const auto& tc_clu1 : cluster1.triggercells()){ 
+            cluTcSet.insert( tc_clu1->detId() );
+            const auto neighbors = triggerGeometry.getNeighborsFromTriggerCell( tc_clu1->detId() );
+            for(const auto neighbor : neighbors){
+                cluTcSet.insert( neighbor );
             }
         }        
             
-        for(unsigned i_clu2(i_clu1+1); i_clu2 < clustersTmp.size(); ++i_clu2){
-
-            const edm::PtrVector<l1t::HGCalTriggerCell> pertinentTC_clu2 = clustersTmp.at(i_clu2).triggercells();
-            for(edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc_clu2 = pertinentTC_clu2.begin(); tc_clu2 != pertinentTC_clu2.end(); ++tc_clu2 ){
-                
-                if( !( cluTcSet.find( (*tc_clu2)->detId() ) == cluTcSet.end() ) && clustersTmp.at(i_clu2).isComplete()==true )
-                {                     
-                    this->mergeClusters( clustersTmp.at(i_clu1), clustersTmp.at(i_clu2) );
-                    clustersTmp.at(i_clu1).setIsComplete(false);
-                }                
-            }                   
+        for(unsigned i_clu2 = i_clu1+1; i_clu2 < clustersTmp.size(); ++i_clu2){
+            l1t::HGCalCluster& cluster2 = clustersTmp.at(i_clu2);
+            // If the cluster has been merged into another one, skip it
+            if(!cluster2.isComplete()) continue;
+            // Check if the TC in clu2 are in clu1 or its neighbors
+            // If yes, merge the second cluster into the first one
+            for(const auto& tc_clu2 : cluster2.triggercells()){ 
+                if( cluTcSet.find(tc_clu2->detId())!=cluTcSet.end() ){
+                    mergeClusters( cluster1, cluster2 );
+                    cluster2.setIsComplete(false);
+                    break;
+                }
+            }
         }
-
     }
 
     /* store clusters in the persistent collection */
-    for( unsigned i(0); i<clustersTmp.size(); ++i ){
-        
+    // only if the cluster contain a TC above the seed threshold
+    for( auto& cluster : clustersTmp ){
         bool saveInCollection(false);
```

Comment:
@mastrolorenzo 
I forgot one thing here: the merged clusters need to be skipped:
```cpp
if(!cluster.isComplete()) continue;
```